File: DiscordConnector/ControlInterface/services/channel_service.py
"""Channel service - business logic for channel operations."""

import logging
from dataclasses import dataclass

logger = logging.getLogger(__name__)


@dataclass
class ChannelData:
    """Domain object for channel."""
    id: int
    name: str
    category_id: int
    position: int


@dataclass
class RoleData:
    """Domain object for role (minimal)."""
    id: int
    name: str
    color: tuple[int, int, int]
    position: int


class ChannelService:
    """Service for channel operations."""

    def __init__(self, controller, db_controller):
        self._ctrl = controller
        self._db_ctrl = db_controller

    async def create_channel(
        self,
        name: str,
        category_id: int | None = None,
        position: int | None = None,
    ) -> ChannelData:
        """Create a new channel and persist to database."""
        channel = await self._ctrl.create_channel(name, category_id, position)

        try:
            await self._db_ctrl.create_channel(
                channel_id=str(channel.id),
                channel_name=channel.name,
                category_id=str(channel.category_id),
            )
        except Exception as e:
            logger.error(f"Failed to save channel to database: {e}")

        return ChannelData(
            id=channel.id,
            name=channel.name,
            category_id=channel.category_id,
            position=channel.position,
        )

    async def delete_channel(self, channel_id: int) -> bool:
        """Delete a channel and remove from database."""
        success = await self._ctrl.delete_channel(channel_id)

        if success:
            try:
                await self._db_ctrl.delete_channel(channel_id=str(channel_id))
            except Exception as e:
                logger.error(f"Failed to delete channel from database: {e}")

        return success

    async def list_channels(self) -> list[ChannelData]:
        """List all channels."""
        channels = await self._ctrl.list_channels()
        return [
            ChannelData(
                id=c.id,
                name=c.name,
                category_id=c.category_id,
                position=c.position,
            )
            for c in channels
        ]

    async def list_channel_roles(self, channel_id: int) -> list[RoleData]:
        """List roles associated with a channel."""
        roles = await self._ctrl.list_channel_roles(channel_id)
        return [
            RoleData(id=r.id, name=r.name, color=r.color, position=r.position)
            for r in roles
        ]

File: DiscordConnector/src/ControlInterface/services/channelService.ts
import type { IDiscordController } from "../../DiscordController";
import type { IDiscordDatabaseController } from "../../DiscordDatabaseController";
import type { ChannelData, RoleData, Snowflake } from "../../types";

export class ChannelService {
  constructor(
    private readonly controller: IDiscordController,
    private readonly dbController: IDiscordDatabaseController,
  ) {}

  async createChannel(name: string, categoryId: Snowflake | null = null, position: number | null = null): Promise<ChannelData> {
    const channel = await this.controller.createChannel(name, categoryId, position);
    if (channel.categoryId !== null) {
      try {
        await this.dbController.createChannel(channel.id, channel.name, channel.categoryId);
      } catch (error) {
        console.error("Failed to save channel to database", error);
      }
    }
    return channel;
  }

  async deleteChannel(channelId: Snowflake): Promise<boolean> {
    const success = await this.controller.deleteChannel(channelId);
    if (success) {
      try {
        await this.dbController.deleteChannel(channelId);
      } catch (error) {
        console.error("Failed to delete channel from database", error);
      }
    }
    return success;
  }

  async listChannels(): Promise<ChannelData[]> {
    return this.controller.listChannels();
  }

  async listChannelRoles(channelId: Snowflake): Promise<RoleData[]> {
    return this.controller.listChannelRoles(channelId);
  }
}
